### DISCUSSION TITLE
Moving primary Morphir development to morphir-dotnet

### DISCUSSION THREAD

**Discussion by DamianReeves** [finos/morphir#165] [Ideas]
Within the Morphir project we have expressed a desire to expand Morphir to other platforms, including:

- JVM/Scala
- .NET
- TypeScript
- Rust
- Python
- And others

There has been quite a bit of discussion around how we can have one central repository from which we target other platforms. The attached PDF represents the proposal I gave for using .NET, F#, and Fable during our Biweekly meeting on [April 6, 2023](https://github.com/finos/morphir/issues/164).

We will use this thread both for discussion and to organize our thoughts on next steps and an action plan to execute on.

[Morphir With Dotnet.pdf](https://github.com/finos/morphir/files/11172405/Morphir.With.Dotnet.pdf)

In order to move the work here we need to port parts of `morphir-elm` but we also will want to add some other things. Here is a list of some things I think we should tackle and/or explore:

| Story | Epic | Category | Description|
|------|------|----------|------------|
| Create an Elm AST/IR| Elm Compatibility | Elm Compatibility| We will need an Elm AST/IR for Morphir backends and frontends but also for us in the Elm Type
| Create an Elm Parser in .NET (write in F#) | Elm FrontEnd | Elm Compatibility| We will need the Elm compiler to use in other Elm compatibility stories|  
| Create a Myriad [^1] Generator that generates F# from Elm| Elm Compatibility | Elm Compatibility | Myriad provides tooling for generating F# code and its use here would be perfect for moving features over |
| Look into a JavaScript to WebAssembly compiler| Elm Compatibility | Elm Compatibility| This would allow us to embed the Elm Frontend's and backends directly in `morphir-dotnet` without needing to rewrite to start, using WASM also means we have no dependency on Node|
| Add Fable 4 as a target for Morphir.IR| Fable Integration | Fable | Fable 4 will allow us to share the IR code across multiple languages|
| Create a DevServer that exposes a REST API | Morphir Develop | Morphir Develop Port/Evolution | We should be able to start to leverage .NET's strong web server stacks to make writing backends easier|
| Create Morphir IR TypeProvider| Morphir IR TypeProvider| Type Providers | With a Morphir IR type provider we can get intellisense and code-gen for Morphir models|
| Add ease of use IR generation features via an F# computation expressopm | Morphir IR| Morphir IR| In F# libraries/tools like Saturn, FsHttp, and others use computation expressions to make a very nice DSL [^2] for builders, adding something like that is not a huge lift and would simplify IR construction|

[^1]: https://moiraesoftware.github.io/myriad/
[^2]: https://sleepyfran.github.io/blog/posts/fsharp/ce-in-fsharp/

**Comment by DamianReeves** [comment]
@stephengoldbaum , @AttilaMihaly , @psmulovics ... have you all had a chance to look at the above. This is far from an exhaustive list.

**Reply by psmulovics** [reply]
I'm at a hackathon yesterday and today, but will try to test this
